searx/searx
# [enh] replace requests by httpx

## What does this PR do?

PR that replace requests by httpx.

`settings.yml` do not require any changes.

### implementation
* use [httpx.AsyncClient](https://www.python-httpx.org/api/#asyncclient) and [uvloop](https://github.com/MagicStack/uvloop)
* the `searx.poolrequests` API are not async.
* use `atexit` module to close HTTP clients (wait at most 3 seconds). Avoid a warning from httpx.

### proxy
* URL http:// supported
* URL https:// not supported
* URL sock4:// , socks5:// , socks5h:// : supported using [httpx-socks](https://pypi.org/project/httpx-socks/)

see `searx.poolrequests.iter_proxies`.

### sources IPs
* supported using a dedicated [transports](https://www.python-httpx.org/advanced/#mounting-transports).
	
### authentification
* see https://www.python-httpx.org/quickstart/#authentication
* `params['auth'] = (user, password)`: basic authentification
* `from httpx import DigestAuth; params['auth'] = DigestAuth(user, password)`: DIGEST authentification

### pool size
* `outgoing.pool_connections` and `outgoing.pool_maxsize` are ìmplementated, see `searx.poolrequests.LIMITS`.
	
### http2:
* disable by default
* can be enabled in settings.yml

### image proxy
* `poolrequests.stream` is an sync function that streams the response from the async httpx client.
* httpx supports sync client but
	* it requires to duplicates the specific implementation of httpx.AsyncHTTPTransport
	* the HTTP2 client may behave differently (according to the issues in encode/httpx)

### requests compatibility
* see https://www.python-httpx.org/compatibility/
* `response.url` is not a string, it is [URL](https://www.python-httpx.org/api/#url):
	* some engines have been modified to use `response.url` instead of `urlparse(response.url)`.
* `response.ok` is the `not response.is_error`.

### performance
* httpx seems faster than requests when http2 is enabled
* httpx is slowed when a lot of data are tranfered (image proxy use case).
* see https://github.com/dalf/pyhttp-benchmark

## Why is this change important?

One step to implement the new architecture.

## How to test this PR locally?

* check these engines:
   * dictzone: check the returned URL
   * elasticsearch: check authentification
   * google: check the catcha detection.
   * seznam: check the catcha detection.
   * stackoverflow: check the catcha detection.
   * yacy: check the authentification.

* check settings.yml:
    * `outgoing.request_timeout`: set a very low limit, and check the timeout
	* `outgoing.pool_maxsize`: set to 1 and send multiple request, check searx is waiting for the connection to be available
	* `outgoing.source_ips`: set multiple IP, check searx uses the different IPs.
    * `outgoing.proxies`: http, socks4, socks5, socks5h: 
		* check searx uses the proxy.
		* check searx throws error when the proxy is not available.
	
* check the image proxy (without morty) works as expected:
    * when the thumbnail is available
	* when the thumbnail is not available (HTTP error).
	* when the thumbnail is bigger than 5MB.
	
* check searx_extra/update/update_external_bangs.py fetchs the external bangs.

* check autocompletion:
	* when there is no error
	* when there is any HTTP error (no Internet connection for example).

* check everything again with http2 enable.

## Author's checklist

`tests/unit/test_poolrequests.py` needs to be updated.

## Related issues

* New version of https://github.com/searx/searx/pull/1856 (which contains general thoughts about HTTP clients, start new discussion here only about httpx).
